design/jira-traceability.md

```diff
@@ -0,0 +1,150 @@
+
+#### Jira Build UI and APIs:
+The  build results are populated in the development section of the work item UI. On further clicking on it, we see the pipeline runs that were associated with the build. 
+
+![Build1](./images/jira-traceability-build-1.png)
+
+
+Drill down view:
+![Build2](./images/jira-traceability-build-2.png)
+
+For the API, the combination of `pipelineId` (pipeline definition identifier) and `buildNumber` (pipeline run identifier) is considered as the unique identifier for a build. 
+
+#### Jira Deployment UI and APIs:
+
+Jira has a separate section for Deployments in its work item UI. In this, only the highest deployed environment category is displayed. Eg. If the work item was deployed to a `test` type environment and  then to `production` type, Jira shows only `production` in deployment section. On clicking on this entry the user can see details of previous deployments.
+
+![Release1](./images/jira-traceability-deploy-1.png)
+
+Drill down view:
+![Release2](./images/jira-traceability-deploy-2.png)
+
+For the API, the combination of `pipelineId` (pipeline definition identifier), `environmentID` (environment identifier), `deploymentSequenceNumber` (deployment identifier) constitute a unique identifier for deployment. 
+
+
+#### When should the data be pushed from Azure Pipelines:?
+ Based on customer interactions, customers prefer seeing the build status as soon as a build completes and deployment status when deployment to an environment is complete. Since build generation and deployments are done at a job level in azure pipelines, we'll push the status information to Jira at job boundaries. For deployments, azure pipelines already has a separate category of jobs, viz, deployment jobs. We'll invoke deployment APIs at the beginning and end of deployment jobs if all the filter conditions are met. For builds, however, we currently do not have a separate category of jobs. For this, we will assume that any job that publishes an artifact is a build job. We'll invoke build API at the end of these jobs if filter conditions are met.  
+
+
+
+## YAML: 
+
+Since traceability primarily involves interacting with external systems, we propose taking configuration for interaction with these external systems in a separate `integrations` section in YAML. As a beginning, we will support a first class integration with JIRA. In future, we may eventually make this section extensible thereby allowing 3rd party authors to create their own integrations. In this document, we will keep focus on JIRA integration. <em>Details on extensibility design are out of scope for this document.</em>
+
+ 
+
+### Schema:
+
+```yaml
+integrations:       
```

Comment:
If you are modelling this as job property, how are we thinking about extensibility? Would we (job execution) handle integrations with every possible artifact? Instead should we have these as tasks (auto injected in the end) which makes it extensible for anyone to author a task that can report back status to their respective work item system? - Model that is similar to download artifacts that we already have?

Reply:
This integration is at a job level, however, future integrations may need to be at a release level/stage level. We would need to evaluate such possibilities before choosing a specific design route like the auto task injection. Since this is an implementation detail, we will take the design discussions separately.